### PR TITLE
Cleanup references on destroy

### PR DESCRIPTION
**Pull request by brandonduff**
Partially addresses this issue: https://github.com/mavenlink/brainstem-js/issues/76

References to associations aren't cleaned up when a model is destroyed. This PR does this without making any extra requests.

### CODE REVIEW

File: src/model.coffee [hunk 1]
```diff
@@ -169,6 +169,24 @@ class Model extends Backbone.Model
...
+        for reference, modelName of collection.modelKlass.associations
+          associationKey = collection.modelKlass.associationDetails(reference).key
+          if _.isArray(modelName) && _.find(modelName, (m) => inflection.singularize(m) == @className())
+            collection.storage.each (model) =>
+              model.set(associationKey, _.without(model.get(associationKey), @id))
+          else if !_.isArray(modelName) && inflection.singularize(modelName) == @className()
+            collection.storage.each (model) =>
+              model.unset(associationKey)
+
+    if(options.wait)
```
Comment (naganowl) [inline src/model.coffee]: Add space 👮

File: src/model.coffee [hunk 2]
```diff
@@ -169,6 +169,24 @@ class Model extends Backbone.Model
       )
       .promise(options.returnValues.jqXhr)
 
+  destroy: (options = {}) ->
+    cleanUpAssociatedReferences = =>
+      for collectionName, collection of @storageManager.collections
+        for reference, modelName of collection.modelKlass.associations
```
Comment (naganowl) [inline src/model.coffee]: Is there a reason we're not using `.each` for these two?

File: src/model.coffee [hunk 3]
```diff
@@ -169,6 +169,24 @@ class Model extends Backbone.Model
...
+          associationKey = collection.modelKlass.associationDetails(reference).key
+          if _.isArray(modelName) && _.find(modelName, (m) => inflection.singularize(m) == @className())
```
Comment (naganowl) [inline src/model.coffee]: For clarity, perhaps `modelName` should be pluralized or named in such a way to reflect it's an array?

File: src/model.coffee [hunk 4]
```diff
@@ -169,6 +169,24 @@ class Model extends Backbone.Model
...
+            collection.storage.each (model) =>
+              model.set(associationKey, _.without(model.get(associationKey), @id))
+          else if !_.isArray(modelName) && inflection.singularize(modelName) == @className()
+            collection.storage.each (model) =>
```
Comment (naganowl) [inline src/model.coffee]: To prevent an extra closure from being created, we can pass `this` as the final argument to `.each`. This would look like so

```
collection.storage.each (model) ->
  ...
, this
```

Comment (naganowl) [inline src/model.coffee]: Also, if you want to make this more Underscore like, you can do

`collection.storage.invoke 'unset', associationKey`

File: src/model.coffee [hunk 5]
```diff
@@ -169,6 +169,24 @@ class Model extends Backbone.Model
...
+              model.unset(associationKey)
+
+    if(options.wait)
+      super(options).then(cleanUpAssociatedReferences)
+    else
+      cleanUpAssociatedReferences()
+      super(options)
```
Comment (naganowl) [inline src/model.coffee]: Instead of this loop, we can rely on the `destroy` event that the model triggers on `destroy` (http://backbonejs.org/docs/backbone.html#section-89). Given that we can convert this entirely into the following

`this.listenTo this, 'destroy', cleanUpAssociatedReferences`

It looks funny, by using `listenTo` allows us to rely on the automatic clean up when views are closed (http://marionettejs.com/docs/v1.8.8/marionette.view.html#view-close)

File: src/model.coffee (hunk 3 above)
Comment (brandonduff) [inline src/model.coffee]: Yeah, we struggled with this because it can be something like `'tasks'` or `['tasks']`. We decided to go with `associator`, since that's what other code in this file referred to it as.

File: src/model.coffee (hunk 4 above)
Comment (brandonduff) [inline src/model.coffee]: Done.

File: src/model.coffee (hunk 1 above)
Comment (brandonduff) [inline src/model.coffee]: Done.

File: src/model.coffee (hunk 5 above)
Comment (brandonduff) [inline src/model.coffee]: Done.

File: package.json [hunk 6]
```diff
@@ -6,7 +6,7 @@
   },
   "filename": "brainstem",
   "standalone": "Brainstem",
-  "version": "0.4.16",
+  "version": "0.4.17-a",
```
Comment (naganowl) [inline package.json]: We should make this non-alpha once this has been completely tested. Can be in a separate PR to make it easier, but you can publish versions outside of merging PRs.

File: src/model.coffee [hunk 7]
```diff
@@ -297,5 +314,11 @@ class Model extends Backbone.Model
   _onAssociatedCollectionChange: (field, collectionChangeDetails) =>
     @attributes[@constructor.associationDetails(field).key] = collectionChangeDetails[1].pluck('id')
 
+  _collectionHasMany: (associator) ->
+    _.isArray(associator) &&
+    _.find(associator, (collectionName) => inflection.singularize(collectionName) == @className())
```
Comment (naganowl) [inline src/model.coffee]: If you wanted to drop this fat arrow, you can pass `this` as the final argument to this method.
